# Task 2 (Detecting depressed subjects)
### Using BoW methods with rolling window approach

In [1]:
P_FILE = "../posts.csv"
TRAIN_TOKEN="../train_df.csv"
TEST_TOKEN="../test_df.csv"
GENERAL_MODELS="../Models"
ROLLING_WINDOW_SIZE=10
TF_IDF_FOUND=True
TASK=2
BASELINE_COMP=False
#if sentiment:
#    MODEL_PATH =f"{GENERAL_MODELS}/BoW/TFIDF_win_{ROLLING_WINDOW_SIZE}_vader_tb" 
#else:
MODEL_PATH =f"{GENERAL_MODELS}/BoW/TFIDF_win_{ROLLING_WINDOW_SIZE}" 

BASIC_PATH=f"{GENERAL_MODELS}/BoW/TFIDF_win_{ROLLING_WINDOW_SIZE}" 

In [2]:
from pathlib import Path
Path(MODEL_PATH).mkdir(parents=True, exist_ok=True)

### opening the preprocessed dataset with pandas

In [3]:
import pandas as pd
import numpy as np
import os
seed=23
np.random.seed(seed)

train_df = pd.read_csv(TRAIN_TOKEN, sep='\t')
test_df = pd.read_csv(TEST_TOKEN, sep='\t')
train_df

User  Post_Nr  \
0       test_subject1345        0   
1       test_subject1345        1   
2       test_subject1345        2   
3       test_subject1345        3   
4       test_subject1345        4   
...                  ...      ...   
174168       subject9959      627   
174169       subject9959      628   
174170       subject9959      629   
174171       subject9959      630   
174172       subject9959      631   

                                                      Raw  \
0                       so many unwanted smith fadeaways.   
1       mid range jumpers hey guys, celtics fan here p...   
2        well he got number tonight so maybe he will b...   
3        i mean he will get pinch hits and an occasion...   
4                   yeah you are probably right. oh well.   
...                                                   ...   
174168             nothing like that clean house feeling    
174169              there is always that one coworker...    
174170  there is always that one coworker you just can...   
174171   that moment when you realize you need a new job    
174172  as an artist, this speaks to me on so many lev...   

                                                  Stemmed  \
0                         so mani unwant smith fadeaway .   
1       mid rang jumper hey guy , celtic fan here pull...   
2       well he got number tonight so mayb he will be ...   
3       i mean he will get pinch hit and an occasion d...   
4                  yeah you are probabl right . oh well .   
...                                                   ...   
174168                     noth like that clean hous feel   
174169                 there is alway that one cowork ...   
174170  there is alway that one cowork you just can no...   
174171     that moment when you realiz you need a new job   
174172  as an artist , this speak to me on so mani lev...   

                                               Lemmatized  Label  
0                      so many unwanted smith fadeaways .      1  
1       mid range jumpers hey guys , celtics fan here ...      1  
2         well he got number tonight so maybe he will ...      1  
3         i mean he will get pinch hits and an occasio...      1  
4                 yeah you are probably right . oh well .      1  
...                                                   ...    ...  
174168              nothing like that clean house feeling      0  
174169              there is always that one coworker ...      0  
174170  there is always that one coworker you just can...      0  
174171    that moment when you realize you need a new job      0  
174172  as an artist , this speaks to me on so many le...      0  

[174173 rows x 6 columns]

In [4]:
def rolling_window(df, window_size,stride, field):
    res_map={}
    for user in df['User'].unique():
        user_df = df[df['User']==user]
        res_map[user]=(user_df['Label'].values[0],{})
        posts = user_df[field].values
        iteration=0
        for i in range(0,len(posts),stride):
            res_map[user][1][iteration]=' '.join((posts[i:i+window_size]))
            iteration+=1
    result_df = pd.DataFrame([(k,k1,v1,v[0]) for k,v in res_map.items() for k1,v1 in v[1].items()], columns = ['User','Window_id','Text','Label'])
    
    return result_df

In [5]:
train_df = rolling_window(train_df,ROLLING_WINDOW_SIZE,1,'Lemmatized')
test_df = rolling_window(test_df,ROLLING_WINDOW_SIZE,1,'Lemmatized')

train_df

User  Window_id  \
0       test_subject1345          0   
1       test_subject1345          1   
2       test_subject1345          2   
3       test_subject1345          3   
4       test_subject1345          4   
...                  ...        ...   
174168       subject9959        627   
174169       subject9959        628   
174170       subject9959        629   
174171       subject9959        630   
174172       subject9959        631   

                                                     Text  Label  
0         so many unwanted smith fadeaways . mid range...      1  
1       mid range jumpers hey guys , celtics fan here ...      1  
2         well he got number tonight so maybe he will ...      1  
3         i mean he will get pinch hits and an occasio...      1  
4         yeah you are probably right . oh well .   i ...      1  
...                                                   ...    ...  
174168  nothing like that clean house feeling there is...      0  
174169  there is always that one coworker ... there is...      0  
174170  there is always that one coworker you just can...      0  
174171  that moment when you realize you need a new jo...      0  
174172  as an artist , this speaks to me on so many le...      0  

[174173 rows x 4 columns]

In [6]:
test_df

User  Window_id  \
0      test_subject1445          0   
1      test_subject1445          1   
2      test_subject1445          2   
3      test_subject1445          3   
4      test_subject1445          4   
...                 ...        ...   
36749       subject9832        995   
36750       subject9832        996   
36751       subject9832        997   
36752       subject9832        998   
36753       subject9832        999   

                                                    Text  Label  
0        yeah he moved across the country , essential...      1  
1        that is true , i never really thought about ...      1  
2        i agree about the relationships evolving par...      1  
3        good point . i just feel like i have screwed...      1  
4        technically , not loving him . but i have be...      1  
...                                                  ...    ...  
36749    exactly scumbag christian girl .. i find mys...      0  
36750  scumbag christian girl .. i find myself doing ...      0  
36751  i find myself doing this everytime i walk thou...      0  
36752    ... snow days keep your children from school...      0  
36753                 why i do not donate to charity ...      0  

[36754 rows x 4 columns]

In [7]:
train_df = pd.concat([train_df,test_df])

In [9]:
"""
if sentiment:
    test_df['TB'] = test_df['Text'].apply(lambda text: TextBlob(text).sentiment)
    test_df['VADER'] = test_df['Text'].apply(lambda text: sia.polarity_scores(text))
    test_df['polarity'] = test_df['TB'].apply(lambda tb: tb[0])
    test_df['subjectivity'] = test_df['TB'].apply(lambda tb: tb[1])
    test_df['negativity'] = test_df['VADER'].apply(lambda v: v['neg'])
    test_df['positivity'] = test_df['VADER'].apply(lambda v: v['pos'])
    test_df['neutrality'] = test_df['VADER'].apply(lambda v: v['neu'])
    test_df['compound'] = test_df['VADER'].apply(lambda v: v['compound'])

    test_df
"""

"\nif sentiment:\n    test_df['TB'] = test_df['Text'].apply(lambda text: TextBlob(text).sentiment)\n    test_df['VADER'] = test_df['Text'].apply(lambda text: sia.polarity_scores(text))\n    test_df['polarity'] = test_df['TB'].apply(lambda tb: tb[0])\n    test_df['subjectivity'] = test_df['TB'].apply(lambda tb: tb[1])\n    test_df['negativity'] = test_df['VADER'].apply(lambda v: v['neg'])\n    test_df['positivity'] = test_df['VADER'].apply(lambda v: v['pos'])\n    test_df['neutrality'] = test_df['VADER'].apply(lambda v: v['neu'])\n    test_df['compound'] = test_df['VADER'].apply(lambda v: v['compound'])\n\n    test_df\n"

In [10]:

train_df = train_df.sample(frac=1, random_state=seed).reset_index(drop=True) 
X_train = train_df['Text']

### Since we are starting the actual training, here are some utility functions which we will use to save our trained models.

In [11]:
import pickle
def save_model(model, name):
    with open(f"{MODEL_PATH}/{name}",'wb') as f:
        pickle.dump( model,f)

### Here we select the stopwords to be considered when building the tf-idf matrix

In [12]:
from nltk.corpus import stopwords
nltk_stop_words = set(stopwords.words("english"))
import gensim
gensim_stop_words = gensim.parsing.preprocessing.STOPWORDS

In [13]:
nltk_stop_words

{'a',
 'about',
 'above',
 'after',
 'again',
 'against',
 'ain',
 'all',
 'am',
 'an',
 'and',
 'any',
 'are',
 'aren',
 "aren't",
 'as',
 'at',
 'be',
 'because',
 'been',
 'before',
 'being',
 'below',
 'between',
 'both',
 'but',
 'by',
 'can',
 'couldn',
 "couldn't",
 'd',
 'did',
 'didn',
 "didn't",
 'do',
 'does',
 'doesn',
 "doesn't",
 'doing',
 'don',
 "don't",
 'down',
 'during',
 'each',
 'few',
 'for',
 'from',
 'further',
 'had',
 'hadn',
 "hadn't",
 'has',
 'hasn',
 "hasn't",
 'have',
 'haven',
 "haven't",
 'having',
 'he',
 'her',
 'here',
 'hers',
 'herself',
 'him',
 'himself',
 'his',
 'how',
 'i',
 'if',
 'in',
 'into',
 'is',
 'isn',
 "isn't",
 'it',
 "it's",
 'its',
 'itself',
 'just',
 'll',
 'm',
 'ma',
 'me',
 'mightn',
 "mightn't",
 'more',
 'most',
 'mustn',
 "mustn't",
 'my',
 'myself',
 'needn',
 "needn't",
 'no',
 'nor',
 'not',
 'now',
 'o',
 'of',
 'off',
 'on',
 'once',
 'only',
 'or',
 'other',
 'our',
 'ours',
 'ourselves',
 'out',
 'over',
 'own',
 'r

In [14]:
gensim_stop_words

frozenset({'a',
           'about',
           'above',
           'across',
           'after',
           'afterwards',
           'again',
           'against',
           'all',
           'almost',
           'alone',
           'along',
           'already',
           'also',
           'although',
           'always',
           'am',
           'among',
           'amongst',
           'amoungst',
           'amount',
           'an',
           'and',
           'another',
           'any',
           'anyhow',
           'anyone',
           'anything',
           'anyway',
           'anywhere',
           'are',
           'around',
           'as',
           'at',
           'back',
           'be',
           'became',
           'because',
           'become',
           'becomes',
           'becoming',
           'been',
           'before',
           'beforehand',
           'behind',
           'being',
           'below',
           'beside',
           'besides'

### Before moving on to the model section, we will assess which parameters to use in the tfidfvectorizer by performing grid search (since it is a fast process with a Naive Bayes algorithm as the final classifier)

In [15]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.naive_bayes import MultinomialNB
from sklearn.pipeline import Pipeline
from sklearn.model_selection import RandomizedSearchCV,GridSearchCV
"""
parameters = {
    'max_features':[5000,7500,10000,12500,15000],
    'max_df': [0.25,0.5,0.75,1.0],
    'stop_words':['english',nltk_stop_words,gensim_stop_words],
    'ngram_range':[(1,1),(1,2)]
}

"""



"\nparameters = {\n    'max_features':[5000,7500,10000,12500,15000],\n    'max_df': [0.25,0.5,0.75,1.0],\n    'stop_words':['english',nltk_stop_words,gensim_stop_words],\n    'ngram_range':[(1,1),(1,2)]\n}\n\n"

In [16]:
from sklearn.model_selection import StratifiedKFold
from sklearn.metrics import f1_score

def tune_vectorizer(df,params, n_folds=5):
    skf = StratifiedKFold(n_splits=n_folds)
    user_label_df =df.drop_duplicates('User')
    users = user_label_df['User'].to_numpy()
    
    labels = user_label_df['Label'].to_numpy()
    model = MultinomialNB()
    f1_scores = []
    for train_index, test_index in skf.split(users, labels):
        transformer = TfidfVectorizer(**params)
        train_users = [users[f] for f in train_index]
        test_users = [users[f] for f in test_index]

        train_folds = df[df['User'].isin(train_users)]
        X_train = transformer.fit_transform(train_folds['Text'])
        test_folds = df[df['User'].isin(test_users)]
        X_test = transformer.transform(test_folds['Text'])
       

        
        
        model.fit(X_train, train_folds['Label'])
        f1_scores.append(f1_score(test_folds['Label'],model.predict(X_test)))

    return f1_scores

In [17]:
sw_lists = {'english':'english','nltk_stop_words':nltk_stop_words,'gensim_stop_words':gensim_stop_words}

In [18]:
def objective(trial):
    parameters = {
        'max_features':trial.suggest_int('max_features',1000,15000,step=1000),
        'max_df': trial.suggest_float('max_df',0.20,1.0,step=0.05),
        'stop_words':trial.suggest_categorical('stop_words',['english','nltk_stop_words','gensim_stop_words']),
        'ngram_range':trial.suggest_int('ngram_range',1,3),
        'sublinear_tf':trial.suggest_int('sublinear_tf',0,1)
        
        
    }
    
  
    
    avg_f1 = train_eval(trial,parameters,5,train_df)
    return np.mean(avg_f1)

In [19]:
from tqdm import tqdm
from sklearn.metrics import f1_score

def train_eval(trial,params, n_folds, df):

    skf = StratifiedKFold(n_splits=n_folds)
    user_label_df =df.drop_duplicates('User')
    users = user_label_df['User'].to_numpy()
    
    labels = user_label_df['Label'].to_numpy()

    #format the vectorizer can accept
    params['stop_words'] = sw_lists[params['stop_words']]
    params['ngram_range'] = (1,params['ngram_range'])
    
    f1_scores = []
    for fold,(train_index, test_index) in enumerate(skf.split(users, labels)):
        transformer = TfidfVectorizer(**params)
        train_users = [users[f] for f in train_index]
        test_users = [users[f] for f in test_index]

        train_folds = df[df['User'].isin(train_users)]
        X_train = transformer.fit_transform(train_folds['Text'])
        test_folds = df[df['User'].isin(test_users)]
        X_test = transformer.transform(test_folds['Text'])
    
        model = MultinomialNB()
        model.fit(X_train, train_folds['Label'])
        f1_scores.append(f1_score(test_folds['Label'],model.predict(X_test)))
    
        trial.report(np.mean(f1_scores), fold)

        if trial.should_prune():
            raise optuna.exceptions.TrialPruned()

    return f1_scores


In [20]:
import optuna
import joblib
if not TF_IDF_FOUND:
    study = optuna.create_study(
        study_name=f"t{TASK}_tfidfvec_{ROLLING_WINDOW_SIZE}", storage=f"mysql://root@localhost:3306/t{TASK}_tfidfvec_{ROLLING_WINDOW_SIZE}",
        load_if_exists=True, direction='maximize')
    joblib.dump(study,f"t{TASK}_tfidfvec_{ROLLING_WINDOW_SIZE}")
    #study = optuna.create_study(study_name=f"tfidfvectorizer_cv_{ROLLING_WINDOW_SIZE}",direction="maximize", sampler=optuna.samplers.TPESampler(), pruner=optuna.pruners.NopPruner())
    #study.optimize(objective, n_trials=30, timeout=(60*60*24))
else:
    study = joblib.load(f"study_cv_tfidfvectorizer_{ROLLING_WINDOW_SIZE}.pkl")

In [21]:
#study = joblib.load("study.pkl")
print("Best trial until now:")
print(" Value: ", study.best_trial.value)
print(" Params: ")
for key, value in study.best_trial.params.items():
    print(f"    {key}: {value}")


Best trial until now:
 Value:  0.668882
 Params: 
    max_df: 0.85
    max_features: 13000
    ngram_range: 2
    stop_words: nltk_stop_words
    sublinear_tf: 1


In [22]:
fig = optuna.visualization.plot_param_importances(study)
fig.write_image(f"tfidf_feature_importance_{ROLLING_WINDOW_SIZE}.png")
fig.show()

In [23]:
fig2=optuna.visualization.plot_parallel_coordinate(study)
fig2.write_image(f"tfidf_parallelc_{ROLLING_WINDOW_SIZE}.png")
fig2.show()

In [24]:
fig3=optuna.visualization.plot_slice(study)
fig3.write_image(f"tfidf_plotslice_{ROLLING_WINDOW_SIZE}.png")
fig3.show()

In [26]:
tfidf_params = study.best_trial.params
tfidf_params['stop_words'] = sw_lists[tfidf_params['stop_words']]
tfidf_params['ngram_range'] = (1,tfidf_params['ngram_range'])


In [27]:
"""
if sentiment:
    from scipy.sparse import hstack
    X_train_transformed = hstack((X_train_transformed,np.array(train_df['polarity'])[:,None]))
    X_train_transformed = hstack((X_train_transformed,np.array(train_df['subjectivity'])[:,None]))
    X_train_transformed = hstack((X_train_transformed,np.array(train_df['negativity'])[:,None]))
    X_train_transformed = hstack((X_train_transformed,np.array(train_df['positivity'])[:,None]))
    X_train_transformed = hstack((X_train_transformed,np.array(train_df['neutrality'])[:,None]))
    X_train_transformed = hstack((X_train_transformed,np.array(train_df['compound'])[:,None]))

    X_test_transformed = hstack((X_test_transformed,np.array(test_df['polarity'])[:,None]))
    X_test_transformed = hstack((X_test_transformed,np.array(test_df['subjectivity'])[:,None]))
    X_test_transformed = hstack((X_test_transformed,np.array(test_df['negativity'])[:,None]))
    X_test_transformed = hstack((X_test_transformed,np.array(test_df['positivity'])[:,None]))
    X_test_transformed = hstack((X_test_transformed,np.array(test_df['neutrality'])[:,None]))
    X_test_transformed = hstack((X_test_transformed,np.array(test_df['compound'])[:,None]))
"""
    



"\nif sentiment:\n    from scipy.sparse import hstack\n    X_train_transformed = hstack((X_train_transformed,np.array(train_df['polarity'])[:,None]))\n    X_train_transformed = hstack((X_train_transformed,np.array(train_df['subjectivity'])[:,None]))\n    X_train_transformed = hstack((X_train_transformed,np.array(train_df['negativity'])[:,None]))\n    X_train_transformed = hstack((X_train_transformed,np.array(train_df['positivity'])[:,None]))\n    X_train_transformed = hstack((X_train_transformed,np.array(train_df['neutrality'])[:,None]))\n    X_train_transformed = hstack((X_train_transformed,np.array(train_df['compound'])[:,None]))\n\n    X_test_transformed = hstack((X_test_transformed,np.array(test_df['polarity'])[:,None]))\n    X_test_transformed = hstack((X_test_transformed,np.array(test_df['subjectivity'])[:,None]))\n    X_test_transformed = hstack((X_test_transformed,np.array(test_df['negativity'])[:,None]))\n    X_test_transformed = hstack((X_test_transformed,np.array(test_df['po

### Now we test the baseline performance of some typical models, using sklearn's default parameters

In [28]:
from sklearn.model_selection import StratifiedKFold
from sklearn.metrics import f1_score
from scipy.sparse import hstack


def custom_cv(model, df, sent =False):
    user_label_df =df.drop_duplicates('User')
    users = user_label_df['User'].to_numpy()
    
    labels = user_label_df['Label'].to_numpy()
    
    skf = StratifiedKFold(n_splits=5)
    f1_scores = []
    for train_index, test_index in skf.split(users, labels):
        vectorizer = TfidfVectorizer(**tfidf_params)

        train_users = [users[f] for f in train_index]
        test_users = [users[f] for f in test_index]
        train_folds = df[df['User'].isin(train_users)].copy()
        X_train = vectorizer.fit_transform(train_folds['Text'])
        test_folds = df[df['User'].isin(test_users)].copy()
        X_test = vectorizer.transform(test_folds['Text'])
        
        if sent:
            X_train = hstack((X_train,np.array(train_folds['polarity'])[:,None]))
            X_train = hstack((X_train,np.array(train_folds['subjectivity'])[:,None]))
            X_train = hstack((X_train,np.array(train_folds['negativity'])[:,None]))
            X_train = hstack((X_train,np.array(train_folds['positivity'])[:,None]))
            X_train = hstack((X_train,np.array(train_folds['neutrality'])[:,None]))
            X_train = hstack((X_train,np.array(train_folds['compound'])[:,None]))

            X_test = hstack((X_test,np.array(test_folds['polarity'])[:,None]))
            X_test = hstack((X_test,np.array(test_folds['subjectivity'])[:,None]))
            X_test = hstack((X_test,np.array(test_folds['negativity'])[:,None]))
            X_test = hstack((X_test,np.array(test_folds['positivity'])[:,None]))
            X_test = hstack((X_test,np.array(test_folds['neutrality'])[:,None]))
            X_test = hstack((X_test,np.array(test_folds['compound'])[:,None]))
            
        

        
        
        model.fit(X_train, train_folds['Label'])
        f1_scores.append(f1_score(test_folds['Label'],model.predict(X_test)))

    return f1_scores

In [29]:
from sklearn.linear_model import SGDClassifier
from sklearn.linear_model import Perceptron
from sklearn.ensemble import ExtraTreesClassifier
if BASELINE_COMP:
        models = {
                'sgdLR':SGDClassifier(random_state=seed,loss='log'),
                'NB':MultinomialNB(),\
                'sgdlSVM':SGDClassifier(random_state=seed,loss='hinge'),
                'ExtraTrees':ExtraTreesClassifier(random_state=seed,n_jobs=-1),\
                'Perceptron':Perceptron(random_state=seed)}
        report=""
        best_model_name = ""
        best_model=None
        best_f1=0
        for model_name, model in models.items():
                res = custom_cv(model,train_df)
                if np.mean(res) > best_f1:
                        best_f1=np.mean(res)
                        best_model_name=model_name
                        best_model=model
                report+=f"{model_name} f1: {round(np.mean(res),3)}\n"
                print(f"{model_name} f1: {round(np.mean(res),3)}")
        #cv
        with open(f"{MODEL_PATH}/cv_baseline_report.txt",'w') as f:
                f.write(report)

In [30]:
if BASELINE_COMP:
    from textblob import TextBlob
    from nltk.sentiment.vader import SentimentIntensityAnalyzer

    sia = SentimentIntensityAnalyzer()
    train_df['TB'] = train_df['Text'].apply(lambda text: TextBlob(text).sentiment)
    train_df['VADER'] = train_df['Text'].apply(lambda text: sia.polarity_scores(text))


    train_df['polarity'] = train_df['TB'].apply(lambda tb: (tb[0]+1/2))
    train_df['subjectivity'] = train_df['TB'].apply(lambda tb: tb[1])
    train_df['negativity'] = train_df['VADER'].apply(lambda v: v['neg'])
    train_df['positivity'] = train_df['VADER'].apply(lambda v: v['pos'])
    train_df['neutrality'] = train_df['VADER'].apply(lambda v: v['neu'])
    train_df['compound'] = train_df['VADER'].apply(lambda v: (v['compound']+1)/2)




    train_df.drop(['VADER','TB'], inplace=True, axis=1)


In [31]:
if BASELINE_COMP:
        res = custom_cv(best_model,train_df, sent=True)
        print(f"{best_model_name} f1: {round(np.mean(res),3)}")
        #ADD CV WHEN CHANGING
        with open(f"{MODEL_PATH}/cv_baseline_report.txt",'a') as f:
                f.write(f"\nBest model with sent:\n{best_model_name} f1: {round(np.mean(res),3)}")

## Hyper parameter tuning

#### Best performing model was NB for window size 10 without SA features, having achieved a 0.669 F1 score

In [33]:
from tqdm import tqdm

def train_eval_tuning(trial,params, df):

    skf = StratifiedKFold(n_splits=5)
    user_label_df =df.drop_duplicates('User')
    users = user_label_df['User'].to_numpy()
    
    labels = user_label_df['Label'].to_numpy()

    
    f1_scores = []
    for fold,(train_index, test_index) in enumerate(skf.split(users, labels)):
        transformer = TfidfVectorizer(**tfidf_params)
        train_users = [users[f] for f in train_index]
        test_users = [users[f] for f in test_index]

        train_folds = df[df['User'].isin(train_users)]
        X_train = transformer.fit_transform(train_folds['Text'])
        test_folds = df[df['User'].isin(test_users)]
        X_test = transformer.transform(test_folds['Text'])
        model = MultinomialNB(**params)
    
        model.fit(X_train, train_folds['Label'])
        f1_scores.append(f1_score(test_folds['Label'],model.predict(X_test)))
    
        trial.report(np.mean(f1_scores), fold)

        if trial.should_prune():
            raise optuna.exceptions.TrialPruned()

    return f1_scores


In [38]:
def tuning_objective(trial):
    parameters = {
        'fit_prior':trial.suggest_int('fit_prior',0,1),
        'alpha': trial.suggest_float('alpha',0.0001,5,log=True),
    }
    
  
    
    avg_f1 = train_eval_tuning(trial,parameters,train_df)
    return np.mean(avg_f1)

In [40]:
study = optuna.create_study(
        study_name=f"t{TASK}_tuning_{ROLLING_WINDOW_SIZE}",
        direction='maximize')
study.optimize(tuning_objective, n_trials=20, timeout=(60*60*24))
joblib.dump(study,f"t{TASK}_tuning_{ROLLING_WINDOW_SIZE}.pkl")
    #study = optuna.create_study(study_name=f"tfidfvectorizer_cv_{ROLLING_WINDOW_SIZE}",direction="maximize", sampler=optuna.samplers.TPESampler(), pruner=optuna.pruners.NopPruner())
study = joblib.load(f"t{TASK}_tuning_{ROLLING_WINDOW_SIZE}.pkl")

[I 2022-06-13 13:40:32,989] A new study created in memory with name: t2_tuning_10
[I 2022-06-13 13:47:14,340] Trial 0 finished with value: 0.6676931236676135 and parameters: {'fit_prior': 0, 'alpha': 0.0010857190400183308}. Best is trial 0 with value: 0.6676931236676135.
[I 2022-06-13 13:53:54,618] Trial 1 finished with value: 0.668994057115821 and parameters: {'fit_prior': 1, 'alpha': 1.2735787969688372}. Best is trial 1 with value: 0.668994057115821.
[I 2022-06-13 14:00:34,286] Trial 2 finished with value: 0.6675329721215091 and parameters: {'fit_prior': 1, 'alpha': 0.0018520506748067405}. Best is trial 1 with value: 0.668994057115821.
[I 2022-06-13 14:07:13,919] Trial 3 finished with value: 0.6683886410009906 and parameters: {'fit_prior': 1, 'alpha': 4.422571135216761}. Best is trial 1 with value: 0.668994057115821.
[I 2022-06-13 14:13:54,173] Trial 4 finished with value: 0.6682531243379622 and parameters: {'fit_prior': 0, 'alpha': 0.2259029484395824}. Best is trial 1 with value: 0.

In [41]:
final_vectorizer = TfidfVectorizer(**tfidf_params)
final_vectorizer.fit(train_df['Text'])
nb_params = study.best_trial.params

final_model = MultinomialNB(**nb_params)
final_model.fit(final_vectorizer.transform(train_df['Text']), train_df['Label'])
save_model(final_vectorizer, "optimized_tfidf.pkl")
save_model(final_model, "optimized_NB.pkl")
